test/flush_worker_test.cc

```diff
@@ -0,0 +1,149 @@
+    }));
+  }
+
+  void expectDispatcherRun() {
+    EXPECT_CALL(*dispatcher_, run(_))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The first dispatcher run is in WorkerImpl::start().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::NonBlock, type);
+        }))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The second dispatcher run is in FlushWorkerImpl::work().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::RunUntilExit, type);
+          simulateTimerLoop();
+        }));
+  }
+
+  void simulateTimerLoop() {
+    int i = 0;
+    while (true) {
+      i++;
+      // At least run the while loop for 100 times.
+      if (i == 100) {
+        promise_.set_value();
+      }
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      if (stopped_) {
+        break;
+      } else if (timer_set_) {
+        timer_set_ = false;
+        timer_cb_();
+      }
+    }
+  }
+
+  NiceMock<Envoy::Api::MockApi> mock_api_;
+  Envoy::Thread::ThreadFactory& thread_factory_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
+  Envoy::Random::RandomGeneratorImpl rand_;
+  NiceMock<Envoy::Event::MockDispatcher>* dispatcher_ = nullptr; // not owned
+  std::unique_ptr<Envoy::Runtime::ScopedLoaderSingleton> loader_;
+  NiceMock<Envoy::LocalInfo::MockLocalInfo> local_info_;
+  NiceMock<Envoy::ProtobufMessage::MockValidationVisitor> validation_visitor_;
+
+  NiceMock<Envoy::Event::MockTimer>* timer_; // not owned
+  Envoy::Event::TimerCb timer_cb_;
+  bool timer_set_{};
+  bool stopped_{false};
+  // Protect stopped_.
+  Envoy::Thread::MutexBasicLockable lock_;
+  std::promise<void> promise_;
+
+  Envoy::Stats::MockSink* sink_ = nullptr; // owned by stats_sinks_
+  std::list<std::unique_ptr<Envoy::Stats::Sink>> stats_sinks_;
+  std::chrono::milliseconds stats_flush_interval_{10};
+};
+
+TEST_F(FlushWorkerTest, WorkerFlushStatsPeriodically) {
+  setupDispatcherTimerEmulation();
+
+  auto worker = std::make_unique<FlushWorkerImpl>(mock_api_, tls_, store_, stats_sinks_,
+                                                  stats_flush_interval_);
+
+  std::thread thread = std::thread([&worker, this] {
+    // Wait for the while loop run at least 100 times in simulateTimerLoop().
+    promise_.get_future().wait();
+    worker->exitDispatcher();
+  });
+
+  expectDispatcherRun();
+  // flush() is called at least 100 times in simulateTimerLoop().
+  EXPECT_CALL(*sink_, flush(_)).Times(testing::AtLeast(100));
+
+  worker->start();
+  worker->waitForCompletion();
+  thread.join();
+  worker->shutdown();
+}
+
+// Verify the final flush is always done in FlushWorkerImpl::shutdownThread()
+// even when the dispatcher and timer is not set up (expectDispatcherRun() is not called).
+TEST_F(FlushWorkerTest, FinalFlush) {
+  auto worker = std::make_unique<FlushWorkerImpl>(mock_api_, tls_, store_, stats_sinks_,
```

Comment:
Why does worker need to be a unique pointer.  Can't it just be a FlushWorkerImpl directly?

Reply:
Done.